src/Microsoft.Health.Fhir.Liquid.Converter/Utilities/ComplexObjectFilterUtility.cs

```diff
@@ -0,0 +1,173 @@
+            if (key.Contains('['))
+            {
+                // Our required path dictates we should have an array and if not
+                // then this is not a match
+                if (input is object[] inputArray)
+                {
+                    // If key is [] then we want to loop over every object in the array
+                    // and check for our future paths
+                    if (key == "[]")
+                    {
+                        if (path.Count == 0)
+                        {
+                            return (value == null) || inputArray.Contains(value);
+                        }
+
+                        foreach (object obj in inputArray)
+                        {
+                            // Clone the queue to scope the path at this level for each loop
+                            var localPath = new Queue<string>(path);
+                            if (ObjHasValueAtPath(obj, localPath, value))
+                            {
+                                return true;
+                            }
+                        }
+
+                        return false;
+                    }
+
+                    // If our array key references a specific index
+                    else
+                    {
+                        // Trim and parse out our numeric index
+                        var indexStr = key.TrimStart('[').TrimEnd(']');
+                        var index = int.Parse(indexStr);
+
+                        if (index >= 0 && inputArray.Count() > index)
+                        {
+                            // Recurse
+                            return ObjHasValueAtPath(inputArray[index], path, value);
+                        }
+                        else
+                        {
+                            // It's not possible for this array to have the element specified
+                            // so this is not a match
+                            return false;
+                        }
+                    }
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            // Our key is referencing an object property
+            else
+            {
+                if (input is Dictionary<string, object> inputObject)
+                {
+                    if (inputObject.ContainsKey(key))
+                    {
+                        // Recurse
+                        return ObjHasValueAtPath(inputObject[key], path, value);
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        // This function splits a string into an array of parts like
+        // `test.path[].to[5].value` -> ['test', 'path', '[]', 'to', '[5]', 'value']
+        public static Queue<string> SplitObjectPath(string path)
+        {
+            var parts = new Queue<string>();
+            var part = new StringBuilder();
```

Comment:
Just as a learning tidbit, a more modern dotnet way of doing this would be to use the [Span<T>](https://docs.microsoft.com/en-us/dotnet/api/system.span-1?view=netcore-3.1) and related types. Reason being that you don't have to create copies of strings, the span is really just a "view" on top of any kind of array that doesn't require copying of data.

Reply:
If you have time for a patch I would be interested in seeing what you mean. From my 10 minutes of reading it seems like a Span would not help with concatenation which is what I am using the StringBuilder for here...